flips/20220927-replacement-for-paths.md

```diff
@@ -0,0 +1,89 @@
+
+The current way of specifying paths is not the best fit for the current use cases:
+
+1. By saving things by specifying storage paths collisions can happen if there is already something stored in that storage path. This will be increasingly more likely as more different dapps exist. This problem is also more prominent in the accounts public domain as each account is free to move things around in their storage without any/much impact to its interaction with the outside world. However in the public domain dapps expect certain capabilities to be at certain paths so that they can interact with them.
+
+2. It is uncommon that the user has multiple things stored with the same type. The current path system does not have any shortcuts for this common case.
+
+
+## Design Proposal
+
+Replace the path addressing `(domain,path)` with `(domain,type,name)` addressing, where the domain is always implicitly know from the method you are using (i.e.: `save` always interacts with storage, `link` always links storage with the public domain). The name is a limited size free text `String` that can be empty and must be unique per type. The Functions getting things from storage would also be split into a **Default** and normal variety (e.i.: `load` -> `loadDefault` and `load`). The empty string name would be a special name, which signifies that if there are multiple objects of the same type take that one when addressing without a name.
+
+```cadence
+// current save function on AuthAccount
+fun save<T>(_ value: T, to: StoragePath)
+
+// current load function on AuthAccount
+fun load<T>(from: StoragePath): T?
+```
+
+```cadence
+// proposed save function on AuthAccount
+fun save<T>(_ value: T, name: String)
+// current load function on AuthAccount
+fun load<T>(name: String): T? {
+    // ... some pseudo-code how loadNamed would work ...
+    // if there are no things of type T stored
+    //     return null
+    // if there is one or multiple things of type T stored
+    //     and if one of them matches name
+    //         return it
+    // otherwise
+    //     return null
+}
+fun loadDefault<T>(): T? {
+    // ... some pseudo-code how load would work ...
+    // if there are no things of type T stored
+    //     return null
+    // if there is only one thing of type T stored
+    //     return it
+    // if there are multiple things of type T stored
+    //     and one of them matches an empty string (the default)
+    //         return it
+    // otherwise
+    //     return null
+}
+```
+
+With this change there is less fear of collision between different daaps (because the types would be different) and the case of having only one thing of a certain type stored is the primary supported case. 
+
+!!TODO!! describe all API changes
```

Comment:
This section will likely need to describe how the changes will affect related APIs, such as storage iteration.